pkg/querier/frontend2/dns_watcher.go

```diff
@@ -0,0 +1,80 @@
+package frontend2
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pkg/errors"
+	"google.golang.org/grpc/naming"
+
+	"github.com/cortexproject/cortex/pkg/util/services"
+)
+
+// Notifications about address resolution. All notifications are sent on the same goroutine.
+type DNSNotifications interface {
+	AddressAdded(address string)
+
+	AddressRemoved(address string)
+}
+
+type dnsWatcher struct {
+	watcher       naming.Watcher //nolint:staticcheck //Skipping for now. If you still see this more than likely issue https://github.com/cortexproject/cortex/issues/2015 has not yet been addressed.
+	notifications DNSNotifications
+}
+
+// NewDNSWatcher creates a new DNS watcher and returns a service that is wrapping it.
+func NewDNSWatcher(address string, dnsLookupPeriod time.Duration, notifications DNSNotifications) (services.Service, error) {
+	resolver, err := naming.NewDNSResolverWithFreq(dnsLookupPeriod)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher, err := resolver.Resolve(address)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &dnsWatcher{
+		watcher:       watcher,
+		notifications: notifications,
+	}
+	return services.NewBasicService(nil, w.watchDNSLoop, nil), nil
+}
+
+// watchDNSLoop watches for changes in DNS and sends notifications.
+func (w *dnsWatcher) watchDNSLoop(servCtx context.Context) error {
+	go func() {
+		// Close the watcher, when this service is asked to stop.
+		// Closing the watcher makes watchDNSLoop exit, since it only iterates on watcher updates, and has no other
+		// way to stop. We cannot close the watcher in `stopping` method, because it is only called *after*
+		// watchDNSLoop exits.
+		<-servCtx.Done()
+		w.watcher.Close()
+	}()
+
+	for {
+		updates, err := w.watcher.Next()
+		if err != nil {
+			// watcher.Next returns error when Close is called, but we call Close when our context is done.
+			// we don't want to report error in that case.
+			if servCtx.Err() != nil {
+				return nil
+			}
+			return errors.Wrapf(err, "error from DNS watcher")
```

Comment:
Documentation for `Next` call says: `It should return an error if and only if Watcher cannot recover.`, so using backoff would not help.

Entire dns_watcher.go is copied and slightly modified from `pkg/querier/frontend/worker.go`, where it works in this way. This is also answer to your question about why it isn't based on Thanos DNS service discovery – I considered that, but decided against the idea in this PR, and just using what we have in `frontend` package already, mostly because then I would be reimplementing the watching service itself, which was not the point of this PR.

In a follow-up PR, my plan is to modify `pkg/querier/frontend/worker.go` to use new DNS watcher. We can then also change DNS watcher to use DNS service discovery from Thanos. I don't have strong opinion on it.

Reply:
Thanks for the detailed explanation. All LGTM!